internal/controllers/virtual_devices_controller.go

```diff
@@ -0,0 +1,233 @@
+	log.Println("avbbbbbbbbb", integration)
+	if err != nil {
+		return helpers.GrpcErrorToFiber(err, "failed to get integration")
+	}
+
+	response := vc.getVirtualDeviceMintPayload(int64(integration.TokenId), vid)
+
+	return c.JSON(response)
+}
+
+// SignVirtualDeviceMintingPayload godoc
+// @Description validate signed signature for vehicle minting
+// @Tags        integrations
+// @Produce     json
+// @Param       integrationNode path int true "token ID"
+// @Param       vehicleID path int true "vehicle ID"
+// @Success     200 {array}
+// @Router      /mint/:integrationNode/:vehicleID [post]
+func (vc *VirtualDeviceController) SignVirtualDeviceMintingPayload(c *fiber.Ctx) error {
+	rawIntegrationNode := c.Params("integrationNode")
+	vehicleNode := c.Params("vehicleID")
+	userID := helpers.GetUserID(c)
+
+	req := &SignVirtualDeviceMintingPayloadRequest{}
+	if err := c.BodyParser(req); err != nil {
+		return fiber.NewError(fiber.StatusBadRequest, "Couldn't parse request.")
+	}
+
+	signature := common.FromHex(req.OwnerSignature)
+	if len(signature) != 65 {
+		return fiber.NewError(fiber.StatusBadRequest, "invalid signature provided")
+	}
+
+	integrationNode, err := strconv.ParseUint(rawIntegrationNode, 10, 64)
+	if err != nil {
+		return fiber.NewError(fiber.StatusBadRequest, "invalid integrationNode provided")
+	}
+
+	user, err := vc.usersClient.GetUser(c.Context(), &pb.GetUserRequest{
+		Id: userID,
+	})
+	if err != nil {
+		return helpers.GrpcErrorToFiber(err, "error occurred when fetching user")
+	}
+
+	vid, err := strconv.ParseInt(vehicleNode, 10, 64)
+	if err != nil {
+		return fiber.NewError(fiber.StatusBadRequest, "invalid vehicle id provided")
+	}
+
+	if err = vc.verifyUserAddressAndNFTExist(c.Context(), user, vid, rawIntegrationNode); err != nil {
+		return err
+	}
+
+	integration, err := vc.deviceDefSvc.GetIntegrationByTokenID(c.Context(), integrationNode)
+	log.Println("dddsfjfsdfj", integration)
+	if err != nil {
+		return helpers.GrpcErrorToFiber(err, "failed to get integration")
+	}
+
+	userAddr := common.HexToAddress(*user.EthereumAddress)
+
+	rawPayload := vc.getVirtualDeviceMintPayload(int64(integration.TokenId), vid)
+
+	hash, _, err := signer.TypedDataAndHash(*rawPayload)
+	if err != nil {
+		vc.log.Err(err).Msg("Error occurred creating has of payload")
+		return fiber.NewError(fiber.StatusBadRequest, "Couldn't verify signature.")
+	}
+
+	signature[64] -= 27
+
+	pub, err := crypto.Ecrecover(hash, signature)
+	if err != nil {
+		vc.log.Err(err).Msg("Error occurred while trying to recover public key from signature")
+		return fiber.NewError(fiber.StatusBadRequest, "Couldn't verify signature.")
+	}
+
+	pubRaw, err := crypto.UnmarshalPubkey(pub)
+	if err != nil {
+		vc.log.Err(err).Msg("Error occurred marshalling recovered public public key")
+		return fiber.NewError(fiber.StatusBadRequest, "Couldn't verify signature.")
+	}
+
+	payloadVerified := bytes.Equal(crypto.PubkeyToAddress(*pubRaw).Bytes(), userAddr.Bytes())
```

Comment:
These are both `common.Address`; you can do `crypto.PubkeyToAddress(*pubRaw) == userAddr`.